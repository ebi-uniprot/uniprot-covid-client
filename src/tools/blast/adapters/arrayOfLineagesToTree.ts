export type TaxNode = {
  name: string;
  count: number;
  children: TaxNode[];
  depth: number;
  id?: number; // optional for now, it'd be good to have it eventually
};

const orderChildren = (node: TaxNode): TaxNode => ({
  ...node,
  children: node.children.map(orderChildren).sort((a, b) => b.count - a.count),
});

const arrayOfLineagesToTree = (lineages?: string[][]) => {
  if (!(lineages && lineages.length)) {
    return null;
  }

  // tree, as a reference to the root of the tree
  const tree: TaxNode = { name: 'All', count: 0, children: [], depth: 0 };

  // looping over all the hits lineages
  for (const lineage of lineages) {
    tree.count += 1;
    // this will be used to navigate the tree
    let currentNode = tree;
    // while loop over all the organism from the left to the right of the array
    while (lineage.length) {
      // extract higher organism to process
      const currentOrganism = lineage.shift();
      if (!currentOrganism) {
        continue; // eslint-disable-line no-continue
      }
      // find it in the current node's children
      let currentOrganismNode = currentNode.children.find(
        (child) => child.name === currentOrganism
      );
      // if we don't have it yet, create it and append it
      if (!currentOrganismNode) {
        currentOrganismNode = {
          name: currentOrganism,
          count: 0,
          children: [],
          depth: currentNode.depth + 1,
        };
        currentNode.children.push(currentOrganismNode);
      }
      currentOrganismNode.count += 1;
      currentNode = currentOrganismNode;
    }
  }
  // here we have the full tree generated
  // now mutate it to order by number of children
  return orderChildren(tree);
};

export default arrayOfLineagesToTree;
