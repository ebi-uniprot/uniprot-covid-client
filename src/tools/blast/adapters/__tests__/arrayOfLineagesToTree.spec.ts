import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

import arrayOfLineagesToTree from '../arrayOfLineagesToTree';

const readFile = promisify(fs.readFile);

let inputParams;
let inputParamsWithTaxons;
let sequence;

beforeAll(async () => {
  inputParams = await readFile(
    path.join(__dirname, '..', '__mocks__', 'input-params.xml'),
    'utf8'
  );
  inputParamsWithTaxons = await readFile(
    path.join(__dirname, '..', '__mocks__', 'input-params-with-taxons.xml'),
    'utf8'
  );
  sequence = await readFile(
    path.join(__dirname, '..', '__mocks__', 'input-sequence.fasta'),
    'utf8'
  );
});

describe('arrayOfLineagesToTree tests', () => {
  it('should handle empty array', () => {
    expect(arrayOfLineagesToTree([])).toBe(null);
  });

  it('should transform an array of lineages to a taxonomy tree', () => {
    expect(
      arrayOfLineagesToTree([
        ['Eukaryota', 'Metazoa', 'Chordata'],
        ['Eukaryota', 'Fungi', 'Dikarya'],
      ])
    ).toMatchSnapshot();
  });

  it('should handle mismatching lineage lengths', () => {
    expect(
      arrayOfLineagesToTree([
        ['Eukaryota', 'Metazoa', 'Chordata'],
        ['Eukaryota', 'Fungi', 'Dikarya'],
        ['Eukaryota', 'Fungi', 'Dikarya', 'Ascomycota', 'Saccharomycotina'],
      ])
    ).toMatchSnapshot();
  });
});
