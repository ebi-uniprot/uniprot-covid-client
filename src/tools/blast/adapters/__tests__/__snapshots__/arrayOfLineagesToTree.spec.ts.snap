// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrayOfLineagesToTree tests should handle mismatching lineage lengths 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "count": 1,
                      "depth": 5,
                      "name": "Saccharomycotina",
                    },
                  ],
                  "count": 1,
                  "depth": 4,
                  "name": "Ascomycota",
                },
              ],
              "count": 2,
              "depth": 3,
              "name": "Dikarya",
            },
          ],
          "count": 2,
          "depth": 2,
          "name": "Fungi",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "count": 1,
              "depth": 3,
              "name": "Chordata",
            },
          ],
          "count": 1,
          "depth": 2,
          "name": "Metazoa",
        },
      ],
      "count": 3,
      "depth": 1,
      "name": "Eukaryota",
    },
  ],
  "count": 3,
  "depth": 0,
  "name": "All",
}
`;

exports[`arrayOfLineagesToTree tests should transform an array of lineages to a taxonomy tree 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "count": 1,
              "depth": 3,
              "name": "Chordata",
            },
          ],
          "count": 1,
          "depth": 2,
          "name": "Metazoa",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "count": 1,
              "depth": 3,
              "name": "Dikarya",
            },
          ],
          "count": 1,
          "depth": 2,
          "name": "Fungi",
        },
      ],
      "count": 2,
      "depth": 1,
      "name": "Eukaryota",
    },
  ],
  "count": 2,
  "depth": 0,
  "name": "All",
}
`;
